pkg/app/mutate_cluster_app.go

```diff
@@ -0,0 +1,159 @@
+package app
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/Masterminds/semver/v3"
+	"github.com/giantswarm/apiextensions-application/api/v1alpha1"
+	"github.com/giantswarm/backoff"
+	"github.com/giantswarm/microerror"
+	releases "github.com/giantswarm/release-operator/v3/api/v1alpha1"
+	"gopkg.in/yaml.v3"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/giantswarm/app-admission-controller/pkg/mutator"
+)
+
+type providerReleaseConfig struct {
+	lastClusterAppVersionWithoutRelease *semver.Version
+}
+
+var providersReleaseConfig = map[string]providerReleaseConfig{
+	"cluster-aws": {
+		lastClusterAppVersionWithoutRelease: semver.New(0, 76, 1, "", ""),
+	},
+}
+
+func (m *Mutator) mutateClusterApp(ctx context.Context, app v1alpha1.App) ([]mutator.PatchOperation, error) {
+	m.logger.Debugf(ctx, "Cluster app mutation for setting App version based on the release. App/Cluster:%s, Namespace:%s\n",
+		app.Name, app.Namespace)
+
+	releaseConfig, providerSupportsReleases := providersReleaseConfig[app.Spec.Name]
+	if !providerSupportsReleases {
+		return nil, nil
+	}
+	lastClusterAppVersionWithoutRelease := releaseConfig.lastClusterAppVersionWithoutRelease
+
+	// Check if app is a cluster-$provider app
+	isClusterApp := (app.Spec.Catalog == "cluster" || app.Spec.Catalog == "cluster-test") && strings.HasPrefix(app.Spec.Name, "cluster-")
+	if !isClusterApp {
+		return nil, nil
+	}
+
+	// Check if the App.spec.version is already set to a cluster-aws release from which we upgrade to new releases
+	if app.Spec.Version != "" {
+		currentClusterAppVersion, err := semver.NewVersion(app.Spec.Version)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+		if !currentClusterAppVersion.GreaterThan(lastClusterAppVersionWithoutRelease) {
+			// This is a cluster that is using an older version of the cluster-$provider app, a version that does
+			// not use new releases, so it does not have release version. Therefore, we don't overwrite cluster app
+			// version here.
+			//
+			// True for cluster-aws versions older or equal to v0.76.1.
+			return nil, nil
+		}
+
+		// Additional temporary check during testing.
+		//
+		// Check if this is a dev build on top of the latest version.
+		isNewDevBuild := currentClusterAppVersion.Major() == lastClusterAppVersionWithoutRelease.Major() &&
```

Comment:
There's already a compare function in semver. It would be worth checking it out:

https://cs.opensource.google/go/x/mod/+/refs/tags/v0.18.0:semver/semver.go;l=111-138

Reply:
That doesn't work for us because we build dev versions incorrectly in the CI.

e.g. if the latest version is `v1.0.0`, the commit on top of that is version like `v1.0.0-566116b9913d6b9a6aa8f455fc1e44e7bffe74b2`, which is older per semver, while it should be newer.

In any case, I will update the check here, as it does not need to check for the version at all, and it is enough to only check if the Helm value `.global.release.version` is set or not (because we set it only for clusters that use new releases). This way the admission controller with the change also works for both old and new clusters.

Reply:
This version check is removed now.